### ISSUE TITLE
ORM does not flush anything when used with Promise.all()

### ISSUE THREAD

**Issue by andrew-sol** [mikro-orm/mikro-orm#2934]
**Describe the bug**
Imagine a simple User entity with just `id` and `username` fields.
We need to save a bunch of users using `Promise.all()`.

> The `users` table should be empty before running the script below.

The script:
```
async run() {
  const users = [
    { username: 'A' },
    { username: 'B' },
    { username: 'C' },
    { username: 'D' },
  ];

  await Promise.all(users.map(async (userData) => this.saveUser(userData)));
}

async saveUser({ username }): Promise<User> {
  const userRepo = getEntityManager().getRepository(User);

  // This query is required. We need another 'await' before 'persistAndFlush' to trigger the bug.
  let user = await userRepo.findOne({ username });

  if (!user) {
    user = new User();
    user.username = username;

    await userRepo.persistAndFlush(user);
  }

  console.log('User ID:', user.id);

  return user;
}
```

Console output:
```
[query] select "u0".* from "users" as "u0" where "u0"."username" = 'A' limit 1 [took 7 ms]
[query] select "u0".* from "users" as "u0" where "u0"."username" = 'C' limit 1 [took 13 ms]
[query] select "u0".* from "users" as "u0" where "u0"."username" = 'B' limit 1 [took 14 ms]
[query] begin
[query] select "u0".* from "users" as "u0" where "u0"."username" = 'D' limit 1 [took 47 ms]
[query] insert into "users" ("created_at", "updated_at", "username") values ('2022-03-21T13:23:01.398Z', '2022-03-21T13:23:01.398Z', 'A') returning "id", "created_at", "updated_at" [took 4 ms]
[query] commit
User ID: undefined
User ID: undefined
User ID: undefined
User ID: 1
```

And in the database we have only 1 user created (user A).

But everything works fine if we remove this line:
```
let user = await userRepo.findOne({ username });
```

**Versions**

| Dependency | Version |
| - | - |
| node | 16.13.2 |
| typescript | - |
| mikro-orm | 5.1.1 |
| your-driver | postgres |

**Comment by B4nan** [comment]
Actually we can make this work if we don't clean up the UoW instance between those "flush steps", will try to handle this case in future version, hopefully it won't break something else (worst case will revert that change). 

Still, if you want to run things in parallel, what I said above is the only way.

**Comment by andrew-sol** [comment]
@B4nan seems like there is another possible pitfall. In my real code I did the same thing as in the example you provided and I ended up with duplicated rows in the database.

On the first iteration `persistAndFlush` inserts a row A.
On the second iteration `persistAndFlush` inserts two rows: A and B (but A was already inserted).
I couldn't reproduce it on that simple example with users though.

But, if I replace this line:
```
const repo = getEntityManager().fork().getRepository(...);
```
with this line:
```
const repo = getMikroORM().em.fork().getRepository(...);
```
then I don't have that duplications.

**Comment by B4nan** [comment]
If you use forks this cannot really happen, as each async handler will have its own context, with clean identity map. Sounds like you are sharing the context somehow, which can lead to undefined behaviour.

**Comment by andrew-sol** [comment]
@B4nan I checked and I don't have any custom context params in the ORM config. I run all this inside a NestJS application (@mikro-orm/nestjs version 4.3.1).

Eventually, I resolved all those issues by replacing `Promise.all()` with a strict promise sequence in a `for` loop. But even with that, I had duplicates until I removed the `.fork()` usage.

I just did a few "turn on random things" tests and found out that it also works fine if I use `Promise.all()` and do the forking this way:
```
const repo = getEntityManager().fork({ useContext: true }).getRepository(...);
```

Tested a couple of times. No accumulating inserts, no duplicates.
